pennylane_sf/gbs.py

```diff
@@ -0,0 +1,246 @@
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+        cutoff_dim (int): Fock-space truncation dimension
+        backend (str): name of the remote backend to be used
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
+        hbar (float): the convention chosen in the canonical commutation
+            relation :math:`[x, p] = i \hbar`
+    """
+    name = "Strawberry Fields variational GBS PennyLane plugin"
+    short_name = "strawberryfields.gbs"
+
+    _operation_map = {
+        "ParamGraphEmbed": GraphEmbed,
+    }
+
+    _observable_map = {
+        "Identity": identity,
+    }
+
+    _circuits = {}
+
+    _capabilities = {"model": "cv", "provides_jacobian": True}
+
+    def __init__(self, wires, *, analytic=True, cutoff_dim, backend="gaussian", shots=1000, hbar=2):
+        if not analytic and backend != "gaussian":
+            raise ValueError("Only the Gaussian backend is supported in non-analytic mode.")
+
+        super().__init__(wires, analytic=analytic, shots=shots, hbar=hbar)
+        self.cutoff = cutoff_dim
+        self.backend = backend
+        self._params = None
+        self._WAW = None
+
+    @staticmethod
+    def _calculate_WAW(params, A, n_mean):
+        """Calculate the :math:`WAW` matrix.
+
+        Rescales :math:`A` so that when encoded in GBS the mean photon number is equal to
+        ``n_mean``.
+
+        Args:
+            params (array[float]): variable parameters
+            A (array[float]): adjacency matrix
+            n_mean (float): mean number of photons
+
+        Returns:
+            array[float]: the :math`WAW` matrix
+        """
+        A *= rescale(A, n_mean)
+        W = np.diag(np.sqrt(params))
+        return W @ A @ W
+
+    @staticmethod
+    def _calculate_n_mean(A):
+        """Calculate the mean number of photons for an adjacency matrix encoded into GBS.
+
+        Args:
+            A (array[float]): adjacency matrix
+
+        Returns:
+            float: mean number of photons
+        """
+        singular_values = np.linalg.svd(A, compute_uv=False)
+        return np.sum(singular_values ** 2 / (1 - singular_values ** 2))
+
+    # pylint: disable=pointless-statement,expression-not-assigned
+    def apply(self, operation, wires, par):
+        self._params, A, _ = par
+
+        if len(self._params) != self.num_wires:
+            raise ValueError(
+                "The number of variable parameters must be equal to the total number " "of wires."
+            )
+
+        self._WAW = self._calculate_WAW(*par)
+        n_mean_WAW = self._calculate_n_mean(self._WAW)
+
+        op = self._operation_map[operation](self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
```

Comment:
I'm almost wondering if the code would be clearer simply doing
```suggestion
        op = GraphEmbed(self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
```

Reply:
Another thought - is there any check anywhere to make sure that only _one_ `ParamGraphEmbed` appears in the circuit?